File: seq_read_bench.cpp
#include <iostream>
#include <cstdlib>
#include <chrono>
#include <string>
#include <format>

#include <fcntl.h>  // For posix open flags
#include <unistd.h> // For posix read

#define FILE_SIZE (1024 * 1024 * 64) // 64 MiB
#define SMALL_CHUNK 100
#define MEDIUM_CHUNK 1024
#define INCREMENTAL 8192
#define INCREMENTAL_START 8192
#define LARGEST_CHUNK (256 * 1024)

struct BenchmarkResult {
    int chunk_size;
    int run_number;
    double read_time_ms;
    double throughput_mbps;
};

void write_result(int data_fd, const struct BenchmarkResult& result) {
    std::string results_str  = std::format("{},{},{:.3f},{:.3f}\n",
        result.chunk_size,
        result.run_number,
        result.read_time_ms,
        result.throughput_mbps
    );
    
    ssize_t written = write(data_fd, results_str.data(), results_str.size());
    if (written != results_str.size()) {
        std::cerr << "Error writing result to file\n";
        std::exit(EXIT_FAILURE);
    }
}

void benchmark_chunk_size(int chunk_size, int data_fd) {
    void* buffer = std::malloc(chunk_size);
    if (buffer == NULL) {
        std::cerr << "Could not allocate chunk buffer!\n";
        exit(1);
    }
    
    for (int run = 0; run < 30; run++) {
        // Open test file
        int test_fd = open("test_file.bin", O_RDONLY);
        if (test_fd == -1) {
            std::cerr << "Could not open test file\n";
            std::free(buffer);
            std::exit(EXIT_FAILURE);
        }
        
        std::cout << "File descriptor is %d\n" << test_fd;
        
        auto start = std::chrono::high_resolution_clock::now();
        
        // Read entire file in chunks
        int total_read = 0;
        ssize_t bytes_read;
        
        while (total_read < FILE_SIZE) {
            int to_read = (FILE_SIZE - total_read > chunk_size) ? 
                         chunk_size : (FILE_SIZE - total_read);
            
            bytes_read = read(test_fd, buffer, to_read);
            if (bytes_read <= 0) break;
            total_read += bytes_read;
        }
        
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> start_stop_diff = end - start;

        // Close test file
        if (close(test_fd) == -1) {
            std::cerr << "Error closing test file\n";
            std::exit(EXIT_FAILURE);
        }
        
        std::cout << "Read " << total_read << "bytes\n";
        
        if (total_read != FILE_SIZE) {
            std::cerr << "Could not read entire file!\n";
            std::free(buffer);
            std::exit(EXIT_FAILURE);
        }
        
        // Calculate metrics
        double read_time_ms = start_stop_diff.count() * 1000.0;
        double throughput_mbps = (FILE_SIZE / (1024.0 * 1024.0)) / start_stop_diff.count();
        
        // Write result
        struct BenchmarkResult result = {chunk_size, run + 1, read_time_ms, throughput_mbps};
        write_result(data_fd, result);
    }
    
    std::free(buffer);
}

int main() {
    // Open data output file
    int data_fd = open("benchmark_results.csv", O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (data_fd == -1) {
        perror("Could not open output file");
        return 1;
    }
    
    // Write CSV header
    std::string header = "chunk_size,run_number,read_time_ms,throughput_mbps\n";
    ssize_t header_written = write(data_fd, header.data(), header.size());
    if (header_written != header.size()) {
        fprintf(stderr, "Error writing CSV header\n");
        close(data_fd);
        return 1;
    }
    
    std::cout << "Starting sequential read benchmark...\n";
    
    // Benchmark small reads (100 bytes)
    std::cout << "Testing small reads (100 bytes)\n";
    benchmark_chunk_size(SMALL_CHUNK, data_fd);
    
    // Benchmark medium reads (1K)
    std::cout << "Testing medium reads (1K)\n";
    benchmark_chunk_size(MEDIUM_CHUNK, data_fd);
    
    // Benchmark large reads (64K)
    std::cout << "Testing incremental reads of 8KiB starting from 8KiB up to 256KiB\n";
    for (int incremental_chunk_size = INCREMENTAL_START; 
        incremental_chunk_size <= LARGEST_CHUNK; 
        incremental_chunk_size +=INCREMENTAL)
    {
        benchmark_chunk_size(incremental_chunk_size, data_fd);
    }  
    
    // Close output file
    if (close(data_fd) == -1) {
        std::cout << "Error closing output file\n";
        return 1;
    }
    
    std::cout << "Benchmark completed. Results saved to benchmark_results.csv\n";
    return 0;
}

File: seq_read_bench.py
#!/usr/bin/env python3
"""
Sequential Read Benchmark Orchestration Script
Creates test files, runs benchmark, and generates performance plots.
"""

import os
import subprocess
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sys
from pathlib import Path

# Configuration
FILE_SIZE_MB = 1
FILE_SIZE_BYTES = FILE_SIZE_MB * 1024 * 1024
TEST_FILE_NAME = "test_file.bin"
RESULTS_FILE = "benchmark_results.csv"
BENCHMARK_ELF = "seq_read_bench.elf.bin"

def create_test_file():
    # Create a 1MiB test file with random data
    print(f"Creating {FILE_SIZE_MB}MiB test file: {TEST_FILE_NAME}")
    
    # Generate random data
    with open(TEST_FILE_NAME, "wb") as f:
        # Write in 64KB chunks to avoid memory issues
        chunk_size = 64 * 1024
        remaining = FILE_SIZE_BYTES
        
        while remaining > 0:
            write_size = min(chunk_size, remaining)
            # Create pseudo-random data (simple pattern to avoid /dev/urandom dependency)
            data = bytearray(write_size)
            for i in range(write_size):
                data[i] = (i * 73 + 17) & 0xFF  # Simple pseudo-random pattern
            f.write(data)
            remaining -= write_size

    print(f"Test file created successfully ({os.path.getsize(TEST_FILE_NAME)} bytes)")

def run_benchmark():
    """Execute the benchmark program."""
    print(f"Running benchmark: boot {BENCHMARK_ELF}")
    
    try:
        # Execute the benchmark
        print("Executing the benchmark 🍾")
        result = subprocess.run(["./seq_read_bench"], timeout=20)

        if result.returncode != 0:
            print(f"Benchmark failed with return code {result.returncode}")
            print(f"STDOUT: {result.stdout}")
            print(f"STDERR: {result.stderr}")
            return False
        
        print("Benchmark completed successfully")
        print(f"Output: {result.stdout}")
        return True
        
    except subprocess.TimeoutExpired:
        print("Benchmark timed out after 5 minutes")
        return False
    except Exception as e:
        print(f"Error running benchmark: {e}")
        return False

def load_and_validate_results():
    """Load benchmark results and validate data."""    
    try:
        df = pd.read_csv(RESULTS_FILE)
        print(f"Loaded {len(df)} benchmark results")
        
        # Validate expected columns
        expected_cols = ['chunk_size', 'run_number', 'read_time_ms', 'throughput_mbps']
        if not all(col in df.columns for col in expected_cols):
            print(f"Missing columns in results file. Expected: {expected_cols}")
            return None
        
        # Validate we have expected chunk sizes and run counts
        chunk_sizes = df['chunk_size'].unique()
        
        print(f"Found chunk sizes: {sorted(chunk_sizes)}")
        
        # Check run counts per chunk size
        for chunk in chunk_sizes:
            run_count = len(df[df['chunk_size'] == chunk])
            print(f"Chunk size {chunk}: {run_count} runs")
        return df
        
    except Exception as e:
        print(f"Error loading results: {e}")
        return None

def generate_plots(df):
    """Generate boxplots for benchmark results."""
    print("Generating performance plots...")
    
    # Set up the plotting style
    plt.style.use('default')
    sns.set_palette("husl")
    
    # Create figure with subplots
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6))
    
    # Prepare data with readable chunk size labels
    df_plot = df.copy()

    chunk_size_map = {
        100: "100B",
        1024: "1KiB"
    }
    for i in range(1, 33):
        size = i * 8192
        chunk_size_map[size] = f"{8*i}KiB"
    expected_chunk_sizes = [100, 1024]
    expected_chunk_sizes.extend([8192 * i for i in range(1, 33)])

    df_plot['chunk_label'] = df_plot['chunk_size'].map(chunk_size_map)
    
    # Plot 1: Read Time
    sns.boxplot(data=df_plot, x='chunk_label', y='read_time_ms', ax=ax1)
    ax1.set_title('Sequential Read Performance - Read Time', fontsize=14, fontweight='bold')
    ax1.set_xlabel('Chunk Size', fontsize=12)
    ax1.set_ylabel('Read Time (ms)', fontsize=12)
    ax1.grid(True, alpha=0.3)
    
    # Add statistics to read time plot
    for i, chunk in enumerate(expected_chunk_sizes):
        chunk_data = df[df['chunk_size'] == chunk]['read_time_ms']
        median_val = chunk_data.median()
        ax1.text(i, median_val, f'Med: {median_val:.1f}ms', 
                ha='center', va='bottom', fontweight='bold', 
                bbox=dict(boxstyle='round,pad=0.3', facecolor='white', alpha=0.8))
    
    # Plot 2: Throughput
    sns.boxplot(data=df_plot, x='chunk_label', y='throughput_mbps', ax=ax2)
    ax2.set_title('Sequential Read Performance - Throughput', fontsize=14, fontweight='bold')
    ax2.set_xlabel('Chunk Size', fontsize=12)
    ax2.set_ylabel('Throughput (MB/s)', fontsize=12)
    ax2.grid(True, alpha=0.3)
    
    # Add statistics to throughput plot
    for i, chunk in enumerate(expected_chunk_sizes):
        chunk_data = df[df['chunk_size'] == chunk]['throughput_mbps']
        median_val = chunk_data.median()
        ax2.text(i, median_val, f'Med: {median_val:.1f}MB/s', 
                ha='center', va='bottom', fontweight='bold',
                bbox=dict(boxstyle='round,pad=0.3', facecolor='white', alpha=0.8))
    
    plt.tight_layout()
    
    # Save the plot
    output_file = "read_benchmark_results.png"
    plt.savefig(output_file, dpi=300, bbox_inches='tight')
    print(f"Plot saved as: {output_file}")
    
    # Display summary statistics
    print("\n=== BENCHMARK SUMMARY ===")
    for chunk in expected_chunk_sizes:
        chunk_name = chunk_size_map[chunk]
        chunk_data = df[df['chunk_size'] == chunk]

        print(f"\n{chunk_name}:")
        print(f"  Read Time - Mean: {chunk_data['read_time_ms'].mean():.2f}ms, "
              f"Median: {chunk_data['read_time_ms'].median():.2f}ms, "
              f"Std: {chunk_data['read_time_ms'].std():.2f}ms")
        print(f"  Throughput - Mean: {chunk_data['throughput_mbps'].mean():.2f}MB/s, "
              f"Median: {chunk_data['throughput_mbps'].median():.2f}MB/s, "
              f"Std: {chunk_data['throughput_mbps'].std():.2f}MB/s")

def main():
    """Main orchestration function."""
    print("=== Sequential Read Benchmark Orchestration ===")

    try:
        # Step 1: Create test file system
        create_test_file()

        # Step 2: Run benchmark
        if not run_benchmark():
            print("Benchmark execution failed!")
            return 1
        
        # Step 3: Load and validate results
        df = load_and_validate_results()
        if df is None:
            print("Failed to load valid results!")
            return 1
        
        # Step 4: Generate plots
        generate_plots(df)
        
        print("\n=== Benchmark completed successfully! ===")
        print(f"Results saved in: fs/{RESULTS_FILE}")
        print(f"Plots saved as: read_benchmark_results.png")
        
        return 0
        
    except KeyboardInterrupt:
        print("\nBenchmark interrupted by user")
        return 1
    except Exception as e:
        print(f"Unexpected error: {e}")
        return 1

if __name__ == "__main__":
    exit_code = main()
    sys.exit(exit_code)
